### PR TITLE
Fingerprint_grow: Trigger on finger scan start and on finger scan misplaced

### PR DESCRIPTION
**Pull request by alexborro**
# What does this implement/fix?

This PR is a follow up of PRs #5365 and #5396 .

As discussed with @jesserockz in PR #5365 , I'm rebasing my original PRs that implements 2 new triggers for the component fingerprint_grow and also fixing a code bug and a generic term introduced by PR #5885 

Description of the new triggers and changes:

- `on_finger_scan_misplaced`: It is triggered only for configurations with sensing_pin. The purpose is to detecting when the user is not properly (entirely) touching the sensor, like just touching it with the finger tip or off center. This way the user might receive, for example, an Aura Led visual indication of the issue.

- `on_finger_scan_start`: It is triggered when the sense_pin detects a finger touching the sensor during the "waiting for a finger" state of the system.

- It includes minor fixes related to the "waiting for finger removal" when sensing_pin is used.

- The dump_config was also improved including the most recent value of the sensors, the sensing_pin configuration and the "System Identifier Code" of the sensor - very likely related to the sensor firmware version.

## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** Implements https://github.com/esphome/feature-requests/issues/1792 and new feature: `on_finger_scan_misplaced` trigger

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3490

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
  on_finger_scan_start:
    - fingerprint_grow.aura_led_control:
        state: ALWAYS_ON
        color: GREEN
        speed: 0
        count: 0

    on_finger_scan_misplaced:
      - homeassistant.event:
          event: esphome.frontdoor_finger_scan_misplaced
      - fingerprint_grow.aura_led_control:
          state: FLASHING
          speed: 25
          color: PURPLE
          count: 2
      - text_sensor.template.publish:
          id: fingerprint_state
          state: "Misplaced finger"
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by alexborro** [comment]
Hi @jesserockz , this is the rebased PR to fix the issues we were discussing in the PR #5365 
I hope that finally I get these features merged.
Cheers.

File: esphome/components/fingerprint_grow/fingerprint_grow.cpp [hunk 1]
```diff
@@ -51,6 +53,7 @@ void FingerprintGrowComponent::update() {
 
 void FingerprintGrowComponent::setup() {
   ESP_LOGCONFIG(TAG, "Setting up Grow Fingerprint Reader...");
+  has_sensing_pin_ = (sensing_pin_ != nullptr);
```
Comment (jesserockz) [inline esphome/components/fingerprint_grow/fingerprint_grow.cpp]:
```suggestion
  this->has_sensing_pin_ = (this->sensing_pin_ != nullptr);
```

File: esphome/components/fingerprint_grow/fingerprint_grow.cpp [hunk 2]
```diff
@@ -15,16 +15,18 @@ void FingerprintGrowComponent::update() {
     return;
   }
 
-  if (this->sensing_pin_ != nullptr) {
-    if (this->sensing_pin_->digital_read()) {
+  if (has_sensing_pin_) {
```
Comment (jesserockz) [inline esphome/components/fingerprint_grow/fingerprint_grow.cpp]:
```suggestion
  if (this->has_sensing_pin_) {
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
I understand adding a misplaced trigger and changing the one invalid call to call the new misplaced trigger instead, but I don't understand why we need to remove the recent invalid trigger entirely that was being triggered on other event types that were previously silently dropped? 
If you are not using that trigger, then there is no harm in it being there, but the other author must have added it to those events for their use case.

**Comment by alexborro** [comment]
I got your point, I will revert back the "invalid" trigger for the cases `IMAGE_MESS` and `INVALID_IMAGE`. In all my tests with this sensor, I never got a situation where the image conversion hit those points, but it indeed doesn't hurt.

By the way, is it a policy to prefix all member variables with `this` keyword, even when it is not shadowed by local variables ? If yes, I apologize and I will prefix all cases of `has_sensing_pin_`
Cheers.

**Comment by alexborro** [comment]
Hi @jesserockz , I reverted back the implementation for the "invalid" trigger and fixed the points you mentioned.
Cheers.

**Comment by jesserockz** [comment]
> By the way, is it a policy to prefix all member variables with this keyword, even when it is not shadowed by local variables ? If yes, I apologize and I will prefix all cases of has_sensing_pin_\nCheers.

https://esphome.io/guides/contributing#codebase-standards

**Comment by alexborro** [comment]
Thanks, I've overlooked this point in the code guidelines (it was 4 months ago when I implemented it), although I followed the remaining because it is natural for me.
Anyway, it is fixed.
Cheers.

**Comment by alexborro** [comment]
Hi @jesserockz  , I made the fixes you requested. Could you please check it out ?

File: esphome/components/fingerprint_grow/fingerprint_grow.cpp [hunk 3]
```diff
@@ -15,16 +15,18 @@ void FingerprintGrowComponent::update() {
     return;
   }
 
-  if (this->sensing_pin_ != nullptr) {
-    if (this->sensing_pin_->digital_read()) {
+  if (this->has_sensing_pin_) {
+    if (sensing_pin_->digital_read()) {
```
Comment (jesserockz) [inline esphome/components/fingerprint_grow/fingerprint_grow.cpp]:
```suggestion
    if (this->sensing_pin_->digital_read()) {
```

File: esphome/components/fingerprint_grow/fingerprint_grow.cpp [hunk 4]
```diff
@@ -15,16 +15,18 @@ void FingerprintGrowComponent::update() {
...
       ESP_LOGV(TAG, "No touch sensing");
       this->waiting_removal_ = false;
       return;
+    } else if (!this->waiting_removal_) {
+      finger_scan_start_callback_.call();
```
Comment (jesserockz) [inline esphome/components/fingerprint_grow/fingerprint_grow.cpp]:
```suggestion
      this->finger_scan_start_callback_.call();
```

File: esphome/components/fingerprint_grow/fingerprint_grow.cpp [hunk 5]
```diff
@@ -122,33 +125,38 @@ void FingerprintGrowComponent::scan_and_match_() {
...
+  uint8_t send_result = this->send_command_();
+  switch (send_result) {
     case OK:
       break;
     case NO_FINGER:
-      if (this->sensing_pin_ != nullptr) {
-        ESP_LOGD(TAG, "No finger");
-        this->finger_scan_invalid_callback_.call();
+      if (this->has_sensing_pin_) {
+        waiting_removal_ = true;
```
Comment (jesserockz) [inline esphome/components/fingerprint_grow/fingerprint_grow.cpp]:
```suggestion
        this->waiting_removal_ = true;
```

File: esphome/components/fingerprint_grow/fingerprint_grow.cpp [hunk 6]
```diff
@@ -122,33 +125,38 @@ void FingerprintGrowComponent::scan_and_match_() {
...
+        ESP_LOGD(TAG, "Finger Misplaced");
+        finger_scan_misplaced_callback_.call();
```
Comment (jesserockz) [inline esphome/components/fingerprint_grow/fingerprint_grow.cpp]:
```suggestion
        this->finger_scan_misplaced_callback_.call();
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/fingerprint_grow/fingerprint_grow.cpp [hunk 7]
```diff
@@ -430,13 +439,21 @@ uint8_t FingerprintGrowComponent::send_command_() {
...
+  LOG_SENSOR("  ", "Status", status_sensor_);
+  ESP_LOGCONFIG(TAG, "    Current Value: %d", (uint8_t) status_sensor_->get_state());
+  LOG_SENSOR("  ", "Capacity", capacity_sensor_);
+  ESP_LOGCONFIG(TAG, "    Current Value: %d", (uint16_t) capacity_sensor_->get_state());
+  LOG_SENSOR("  ", "Security Level", security_level_sensor_);
+  ESP_LOGCONFIG(TAG, "    Current Value: %d", (uint8_t) security_level_sensor_->get_state());
+  LOG_SENSOR("  ", "Last Finger ID", last_finger_id_sensor_);
+  ESP_LOGCONFIG(TAG, "    Current Value: %d", (uint32_t) last_finger_id_sensor_->get_state());
+  LOG_SENSOR("  ", "Last Confidence", last_confidence_sensor_);
+  ESP_LOGCONFIG(TAG, "    Current Value: %d", (uint32_t) last_confidence_sensor_->get_state());
```
Comment (jesserockz) [inline esphome/components/fingerprint_grow/fingerprint_grow.cpp]: There are a few more `this->` missing here and GitHub wont let me add review suggestions.

**Comment by alexborro** [comment]
Hi @jesserockz , this is embarrassing, but I think I applied the "this" keyword to only one of my branches, not to the other. Thanks for pointing out my mistake and I apologize for taking your time on this.
Now I triple checked and all class members have the keyword.

**Review by jesserockz — APPROVED** [review]
Thanks for putting up with the requests =)

**Pull request merged**